tools/integration_tests/operations/list_dir_test.go

```diff
@@ -0,0 +1,208 @@
+
+	"github.com/googlecloudplatform/gcsfuse/tools/integration_tests/util/setup"
+)
+
+func getDirNameFromPath(dirPath string) (dirName string) {
+	if strings.Contains(dirPath, EmptySubDirInDirectoryForListTest) {
+		// testBucket/directoryForListTest/emptySubDirectoryForListTest
+		dirName = EmptySubDirInDirectoryForListTest
+	} else if strings.Contains(dirPath, SecondSubDirectoryForListTest) {
+		// testBucket/directoryForListTest/secondSubDirectoryForListTest
+		dirName = SecondSubDirectoryForListTest
+	} else if strings.Contains(dirPath, FirstSubDirectoryForListTest) {
+		// testBucket/directoryForListTest/firstSubDirectoryForListTest
+		dirName = FirstSubDirectoryForListTest
+	} else if strings.Contains(dirPath, DirectoryForListTest) {
+		// testBucket/directoryForListTest
+		dirName = DirectoryForListTest
+	} else if strings.Contains(dirPath, setup.MntDir()) {
+		// testBucket/
+		dirName = setup.MntDir()
+	}
+
+	return dirName
+}
+
+func createDirectoryWithFile(dirPath string, filePath string, t *testing.T) {
+	err := os.Mkdir(dirPath, setup.FilePermission_0600)
+	if err != nil {
+		t.Errorf("Mkdir at %q: %v", dirPath, err)
+		return
+	}
+
+	_, err = os.Create(filePath)
+	if err != nil {
+		t.Errorf("Error in creating file %v:", err)
+	}
+}
+
+func checkIfListedCorrectDirectory(dirPath string, obj fs.DirEntry, t *testing.T) {
+	dirName := getDirNameFromPath(dirPath)
+
+	switch dirName {
+	case setup.MntDir():
+		{
+			// testBucket/directoryForListTest    -- Dir
+			if obj.Name() != DirectoryForListTest || obj.IsDir() != true {
+				t.Errorf("Listed incorrect object.")
+			}
+		}
+	case DirectoryForListTest:
+		{
+			// testBucket/directoryForListTest/fileInDirectoryForListTest     -- File
+			// testBucket/directoryForListTest/firstSubDirectoryForListTest   -- Dir
+			// testBucket/directoryForListTest/secondSubDirectoryForListTest  -- Dir
+			// testBucket/directoryForListTest/emptySubDirectoryForListTest   -- Dir
+			if (obj.Name() != FileInDirectoryForListTest && obj.IsDir() == true) && (obj.Name() != FirstSubDirectoryForListTest && obj.IsDir() != true) && (obj.Name() != SecondSubDirectoryForListTest && obj.IsDir() != true) && (obj.Name() != EmptySubDirInDirectoryForListTest && obj.IsDir() != true) {
+				t.Errorf("Listed incorrect object")
+			}
+		}
+	case FirstSubDirectoryForListTest:
+		{
+			// testBucket/directoryForListTest/firstSubDirectoryForListTest/fileInFirstSubDirectoryForListTest     -- File
+			if obj.Name() != FileInFirstSubDirectoryForListTest && obj.IsDir() == true {
+				t.Errorf("Listed incorrect object")
+			}
+		}
+	case SecondSubDirectoryForListTest:
+		{
+			// testBucket/directoryForListTest/secondSubDirectoryForListTest/fileInSecondSubDirectoryForListTest   -- File
+			if obj.Name() != FileInSecondSubDirectoryForListTest && obj.IsDir() == true {
+				t.Errorf("Listed incorrect object")
+			}
+		}
+	}
+}
+
+func checkIfListedDirectoryHasCorrectNumberOfObjects(dirPath string, numberOfObjects int, t *testing.T) {
```

Comment:
It is getting a little complicated. Instead of creating functions like getDirNameFromPath, checkIfListedDirectoryHasCorrectNumberOfObjects etc, can we keep it simple i.e. a function to create the directory structure (with files) and then list and check if results are expected in TestListDirectoryRecursively (remove recursive from function name) ?

Reply:
Do you mean
Create directory structure.
then list directory one by one and check in the test it self if directory structure is matching or not ?

If yes then this will not be a recursive test right as we added in our test design?

Reply:
Yes,

for recursively listing a directory, you can check https://pkg.go.dev/path/filepath#Walk or similar